### PR TITLE
Refactor permutation

### PR DESCRIPTION
**Pull request by FarnazH**
I built upon @Ali-Tehrani's refactoring of `permutation_2sided` in this PR. This is mostly adding more comments and minor refactoring to make the code even more clear. I am making a major refactoring of the `permutation_2sided` function, based on conversations with @PaulWAyers & @Ali-Tehrani, but I wanted to share this first.

### CODE REVIEW

**Comment by FarnazH** [comment]
@fwmeng88, @Ali-Tehrani, and @PaulWAyers: I worked on the (major) refactoring of `permutation_2sided`, and to make these changes be finalized faster, I am pushing those comments here for review (instead of waiting for this PR to be reviewed/merged before sharing the new refactoring).

File: procrustes/permutation.py [hunk 1]
```diff
@@ -132,92 +132,69 @@ def permutation(
...
-        matching algorithm from [1] is applied.
-        If false, the two permutation matrices can be different, i.e.
-        it is the two-sided permutation Procrustes with two transformations
-        (known as the regular two-sided permutation Procrustes here). An flip-flop
-        algorithm taken from [2] is used to iteratively solve this problem. Global optima is not
-        guaranteed here and may need to do an additional k-opt search.
-        Default=True.
+        If `True`, the single-transformation Procrustes is performed to obtain :math:`\mathbf{P}`.
+        If `False`, the two-transformations Procrustes is performed to obtain :math:`\mathbf{P}_1`
+        and :math:`\mathbf{P}_2`.
```
Comment (fwmeng88) [inline procrustes/permutation.py:171]: We should add **optional** after single,

```python
single : bool, optional
    If `True`, the single-transformation Procrustes is performed to obtain :math:`\mathbf{P}`...
```

File: procrustes/permutation.py [hunk 2]
```diff
@@ -132,92 +132,69 @@ def permutation(
...
+    method : str, optional
+        The method to solve for permutation matrices. For `single=False`, these include "flip-flop"
+        and "k-opt" methods. For `single=True`, these include "approx-normal1", "approx-normal2",
+        "approx-umeyama", "approx-umeyama-svd", "k-opt", "soft-assign", and "nmf".
```
Comment (fwmeng88) [inline procrustes/permutation.py:175]: Do we need to add detailed explanations for all the possible arguments, either here or in the manuscript? For example, `nmf` stands for non-negative matrix factorization, but I would assume someone would have to spend some time digging out what is `nmf` and how it works.

**Review by fwmeng88 — CHANGES_REQUESTED** [review]
Thank you so much for this big improvement!
I only two minor comments.

File: procrustes/permutation.py (hunk 2 above)
Comment (FarnazH) [reply procrustes/permutation.py:175]: Yes, the docstring needs updates and extensions (for both arguments & notes sections). I didn't want to do it before you all agree on the refactoring. So, I will do it as the last step.

File: procrustes/permutation.py (hunk 1 above)
Comment (FarnazH) [reply procrustes/permutation.py]: done. thanks.

**Comment by FarnazH** [comment]
@fwmeng88, @Ali-Tehrani, and @PaulWAyers: The refactored `permutation_2sided` is ready for review. The goal was to make this function more flexible. The current implementation allows one to compute approximate permutation methods as easily as more expensive methods. It also makes it easy to add new algorithms. **Please be critical in your review** :-)

The docstring needs to be updated (for both arguments & notes) to better describe the code functionality, but I didn't want to do it until we all agree on the API changes, etc. In the meanwhile, I am double-checking the implemented algorithms against the paper's appendix to fix any discrepancies.

**Review by fwmeng88 — APPROVED** [review]
Thank you so much for moving our package on! @FarnazH 

I am good with this PR now.

**Comment by Ali-Tehrani** [comment]
I haven't found any logical bugs, and this improved code is really well-done.

**Comment by Ali-Tehrani** [comment]
For the notes, we could do something similar to [SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize).  Below is a template.

Examples
-------------
Do an example using one of the umeyama, then using nmf.

See Also
------------
kopt_heuristic_single 
    Search for k-fold permutations for the single transformation of any function.
kopt_heuristic_double
    Search for all k-fold permutations for double transformation of any function.
softassign
    Optimize the generalized quadratic assignment problem.

Notes
--------
Explain single (one-sided) and double (two-sided) permutation Procrustes problem.

**Single Transformation**

The  method "approx-normal1" found in [This paper] is @PaulWAyers.  This is useful when one wants a decent initial guess.

The method "approx-normal2" found in [This paper] is  @PaulWAyers ... This is useful when one wants a decent initial guess.

The method "approx-umeyama" found in [3] is computes the closest permutation matrix to :math:`|U_A| |U_B|.T`, where U_A, U_B are the eigenvectors of A, B, respectively and |A| is the matrix with absolute value applied to each matrix entry.  This is useful when one wants a decent initial guess.

The  method "approx-umeyama-svd" is a variate of "approx-umeyama" computes the closest permutation matrix to :math:`|U| |V|.T`, where U,V are the left and right unitary matrices of the singular value decomposition of :math:`|U_A| |U_B|.T`.  This is useful when one wants a decent initial guess.

The method "nmf" found in [1] is an iterative fixed-point, gradient-based method that updates the permutation starting with 'guess_p2'.  This is the recommended method of choice with providing initial guess from the methods above.

The method  "k-opt" is a greedy, local search that searches over all k-fold permutations of the initial guess `guess_p2`.  This search is slow when kopt is large. All permutations are searched when kopt is equal to the rows of :math:`A`.    This is useful to refine the results from the other algorithms.

The method "soft-assign" in [4]  is a gradient-based iterative method that solves the permutation problem over doubly stochastic matrices, then finds the closest permutation matrix.  The initial guess `guess_p2` can be provided to this algorithm.  This is useful when other methods are slow or provide poor results.

**Double Transformation**

The method "flip-flop" based in [2] is an iterative method that updates :math:`P_1, P_2` while keeping the other fixed starting with  the initial guesses `guess_p1` and `guess_p2`.  This is done by solving the one-sided permutation problem.
 
The method  "k-opt" is a greedy, local search that searches over all k-fold permutations of the rows of `guess_p1` and all k-fold permutation of the rows of `guess_p2`.  This search is slow when kopt is large.  This is useful to refine the results from the other algorithms.

References
---------------
...
[4] Rangarajan, A., Yuille, A. L., Gold, S., & Mjolsness, E. (1997). A convergence proof for
       the softassign quadratic assignment algorithm. Advances in neural information processing
        systems, 620-626.
[5] This paper

**Review by Ali-Tehrani — APPROVED** [review]

**Comment by FarnazH** [comment]
Thanks for the reviews, and thanks @Ali-Tehrani for the docstring template; that is very useful/helpful. Just not to hold things back any longer, I think we can go on and merge this PR, so @Ali-Tehrani can implement the `soft-assign` method. I will take care of the docstring in another PR. I will let you merge @fwmeng88 wherever you are ready.

**Comment by fwmeng88** [comment]
Thanks for the big improvement! I am going to merge this PR.

**Pull request merged**